File: 07 hashtables/hashtable.js
class Hashtable {
  constructor(size) {
    this.data = new Array(size);
  }

  _hash(key) {
    let hash = 0;

    for (let i = 0; i < key.length; i++) {
      hash = (hash + key.charCodeAt(i) * i) % this.data.length;
    }

    return hash;
  }

  set(key, value) {
    let address = this._hash(key);
    if (!this.data[address]) {
      this.data[address] = [];
    }

    this.data[address].push([key, value]);
  }


  get(key) {
    let address = this._hash(key);

    let item = this.data[address];
    if (undefined == item)
      return undefined;

    for (let i = 0; i < item.length; i++) {
      if (item[i][0] == key)
        return item[i][1];
    }

    return undefined;
  }

  keys() {
    const keysArr = [];

    for (let i = 0; i < this.data.length; i++) {
      itm = this.data[i];

      if (!itm)
        continue;

      for (let j = 0; j < itm.length; j++) {
        if (keysArr.some((a) => {
          return a == itm[j][0];
        }) == false)
          keysArr.push(itm[j][0]);
      }
    }

    return keysArr;
  }
  remove(key) {
    let address = this._hash(key);
    this.data[address] = undefined;
  }
}

File: 07 hashtables/hashtable.py
class Hashtable(object):
  def __init__(self, size):
    self.data = [None] * size
  
  def __hash(self, key):
    hash = 0

    for  i in range (len(key)):
      hash = (hash + ord(key[i]) * i) % len(self.data)

    return hash

  def set(self, key, value):
    address = self.__hash(key)

    if (self.data[address] is None):
      self.data[address] = []

    self.data[address].append([key, value])

    print (self.data)

  def get(self, key):
    address = self.__hash(key)

    item = self.data[address]
    if (item is None):
      return None

    for i in range(len(item)):
      if (item[i][0] == key):
        return item[i][1];

    return None

  def keys(self):
    keysArr = []

    for i in range(len(self.data)):
      item = self.data[i]
      if (item is None):
        continue

      for j in range(len(item)):
        if not item[j][0] in keysArr:
          keysArr.append(item[j][0]);

    return keysArr
